tests/functional/test_new_resolver.py

```diff
@@ -1278,3 +1298,534 @@ def test_new_resolver_no_fetch_no_satisfying(script):
+    script, make_fake_wheel
+):
+    initial_path = make_fake_wheel("base", "0.1.0", "fakepy1-fakeabi-fakeplat")
+
+    constrained = script.scratch_path / "constrained"
+    constrained.mkdir()
+
+    final_path = constrained / initial_path.name
+
+    initial_path.rename(final_path)
+
+    constraints_file = script.scratch_path / "constraints.txt"
+    constraints_file.write_text("base @ " + path_to_url(final_path))
+
+    result = script.pip(
+        "install",
+        "--implementation", "fakepy",
+        '--only-binary=:all:',
+        "--python-version", "1",
+        "--abi", "fakeabi",
+        "--platform", "fakeplat",
+        "--target", script.scratch_path / "target",
+        "--no-cache-dir", "--no-index",
+        "-c", constraints_file,
+        "base",
+    )
+
+    dist_info = Path("scratch", "target", "base-0.1.0.dist-info")
+    result.did_create(dist_info)
+
+
+def test_new_resolver_handles_incompatible_wheel_tags_in_constraint_url(
+    script, make_fake_wheel
+):
+    initial_path = make_fake_wheel("base", "0.1.0", "fakepy1-fakeabi-fakeplat")
+
+    constrained = script.scratch_path / "constrained"
+    constrained.mkdir()
+
+    final_path = constrained / initial_path.name
+
+    initial_path.rename(final_path)
+
+    constraints_file = script.scratch_path / "constraints.txt"
+    constraints_file.write_text("base @ " + path_to_url(final_path))
+
+    result = script.pip(
+        "install",
+        "--no-cache-dir", "--no-index",
+        "-c", constraints_file,
+        "base",
+        expect_error=True,
+    )
+
+    assert (
+        "Cannot install base because these package versions have conflicting "
+        "dependencies."
+    ) in result.stderr, str(result)
+
+    assert_not_installed(script, "base")
+
+
+def test_new_resolver_avoids_incompatible_wheel_tags_in_constraint_url(
+    script, make_fake_wheel
+):
+    initial_path = make_fake_wheel("dep", "0.1.0", "fakepy1-fakeabi-fakeplat")
+
+    constrained = script.scratch_path / "constrained"
+    constrained.mkdir()
+
+    final_path = constrained / initial_path.name
+
+    initial_path.rename(final_path)
+
+    constraints_file = script.scratch_path / "constraints.txt"
+    constraints_file.write_text("dep @ " + path_to_url(final_path))
+
+    index = script.scratch_path / "index"
+    index.mkdir()
+
+    index_dep = create_basic_wheel_for_package(script, "dep", "0.2.0")
+
+    base = create_basic_wheel_for_package(
+        script, "base", "0.1.0"
+    )
+    base_2 = create_basic_wheel_for_package(
+        script, "base", "0.2.0", depends=["dep"]
+    )
+
+    index_dep.rename(index / index_dep.name)
+    base.rename(index / base.name)
+    base_2.rename(index / base_2.name)
+
+    script.pip(
+        "install",
+        "--no-cache-dir", "--no-index",
+        "-c", constraints_file,
+        "--find-links", script.scratch_path / "index",
+        "base",
+    )
+
+    assert_installed(script, base="0.1.0")
+    assert_not_installed(script, "dep")
```

Comment:
Can some of these tests be logically combined with `@pytest.mark.parametrize()`? There seems to be a lot of duplicate setups and I’m struggling to identify their differences.

Reply:
I just pushed what I got done so far. I assume there's more work to be done, since I only got the things that were very, very obvious to me.